MSC4388: Secure out-of-band channel for sign in with QR

hughns opened the pull request:
[Rendered](https://github.com/matrix-org/matrix-spec-proposals/blob/element-hq/oidc-qr-secure-channel/proposals/4388-secure-qr-channel.md)

This was originally part of MSC4108 but has been split out to make review easier.

## Implementations

- Rendezvous session
  - [x] Synapse rendezvous session API:
    - https://github.com/element-hq/synapse/pull/19127
    - https://github.com/element-hq/synapse/pull/19573
  - [x] Ruma typings:
    - https://github.com/ruma/ruma/pull/2340
    - https://github.com/ruma/ruma/pull/2419
  - [x] JS SDK: https://github.com/matrix-org/matrix-js-sdk/pull/5219
- QR encoding/decoding
  - [x] Rust SDK: https://github.com/matrix-org/matrix-rust-sdk/pull/6089 
- Secure channel
  - [x] vodozemac: https://github.com/matrix-org/vodozemac/pull/289
  - [x] Rust SDK: https://github.com/matrix-org/matrix-rust-sdk/pull/6173
  - [x] Web Crypto bindings:
    - https://github.com/matrix-org/matrix-sdk-crypto-wasm/pull/290
    - https://github.com/matrix-org/matrix-sdk-crypto-wasm/pull/310
  - [x] JS SDK: https://github.com/matrix-org/matrix-js-sdk/pull/5219
  - [x] Element Web: (using JS SDK)
    - https://github.com/element-hq/element-web/pull/32799
  - [x] Element X iOS & Android: (using Rust SDK)

## To-dos

- [x] Implementations (see above)
- [x] Review `Sec-Fetch-*` requirements
- [x] Minor: Update matrix-rust-sdk with the revised AAD spec from 5b6dc1eded7aa66cbd465469ad8a5b19fa6b883b

----

SCT Stuff:

[MSC Checklist](https://github.com/matrix-org/matrix-spec-proposals/pull/4388#issuecomment-4486946848)

[FCP tickyboxes](https://github.com/matrix-org/matrix-spec-proposals/pull/4388#issuecomment-4486952984)

Designated reviewers:

* [x] @erikjohnston 
* [x] @richvdh

hughns committed:
Secure out-of-band channel for sign in with QR

hughns committed:
MSC4388

hughns committed:
Overview diagram

hughns committed:
Move example API usage back to after the API definitions

poljar commented on proposals/4388-secure-qr-channel.md:
```suggestion
they are both in the correct state to the user performing the QR code login process.

Additionally the HPKE RFC exclusively defines unidirectional encryption. [Oblivious HTTP](https://www.rfc-editor.org/rfc/rfc9458) is used to enable bidirectionl encryption in our chosen HPKE scheme.

```

poljar committed:
Switch from ECIES to HPKE

hughns commented on proposals/4388-secure-qr-channel.md:
```suggestion
Bidirectional encryption uses a separate encryption key for each sender, both derived from a shared secret using HKDF.
```

Perhaps?

hughns commented on proposals/4388-secure-qr-channel.md:
```suggestion
TaggedCiphertext := Context_S.Seal("MATRIX_QR_CODE_LOGIN_INITIATE", "")
```

hughns reviewed (commented):
@poljar I've reviewed the HPKE changes alongside the RFCs and have some questions/things needing clarification.

hughns commented on proposals/4388-secure-qr-channel.md:
Hmm, is it the nonce that is incremented? Or is the sequence number that is incremented (which is then XORed)?

Speaking of which, I don't see when the aforementioned XORing happens?

hughns replied:
Ah ha. I get it now.

I don't think we need to repeat the definition, but instead directly refer to it. I've attempted to do so in 98f94a02822d8ac20ce0b7f05e7cfde09289c891.

poljar commented on proposals/4388-secure-qr-channel.md:
Yes, it's the same as `concat()` in the HPKE RFC. Oh I thought I used `||` exclusively since it's what we used in the previous versions of this MSC, but a bunch of `concat()` call snuck in.

We can switch to `concat()` if you think that's better or we can revert to `||` in the places where I left the `concat()` calls.

poljar commented on proposals/4388-secure-qr-channel.md:
Oh indeed, the parenthesis should include the nonce as well.

hughns committed:
Standardise on || for concatenation

hughns committed:
Fix encoding of LoginOkMessage

hughns commented on proposals/4388-secure-qr-channel.md:
Ditto:

```suggestion
CheckBytes := Context_G.Export("MATRIX_QR_CODE_LOGIN_CHECKCODE" || Gp || Sp , size=2)
```

hughns committed:
Refer to nonce logic described in HPKE

hughns commented on proposals/4388-secure-qr-channel.md:
```suggestion
LoginOkMessage := UnpaddedBase64Encode(TaggedCiphertext || ResponseBaseNonce)
```

hughns commented on proposals/4388-secure-qr-channel.md:
```suggestion
Salt := Sp || ResponseBaseNonce
```

poljar commented on proposals/4388-secure-qr-channel.md:
I think that's a term that's somewhat commonly used when different sources of randomness are combined.

But I think "combined" works better here. Agree.

poljar commented on proposals/4388-secure-qr-channel.md:
Makes sense.

poljar commented on proposals/4388-secure-qr-channel.md:
Sure, I don't mind them being gone. Not sure if @dkasak has some attachment to those separators.

hughns committed:
Clean up HPKE section

hughns committed:
Update proposals/4388-secure-qr-channel.md

hughns committed:
Remove | from concatenated strings

hughns committed:
Change expires_ts to expires_in_ms

hughns committed:
Use binary packing for LoginInitiateMEssage and LoginOkMessage

hughns committed:
Fix unstable rendezvious endpoint path

hughns committed:
Use HPKE functions directly where possible and don't repeat what is in the RFCs

reivilibre commented on proposals/4388-secure-qr-channel.md:
```diff
@@ -0,0 +1,1014 @@
+This proposal forms part of [MSC4108] to make it easy to sign in on a new device with the help of an existing device.
+
+It proposes a mechanism for a new an existing device to establish a secure out-of-band channel through which they can
+communicate to facilitate the sign in on the new device.
```
```suggestion
It proposes a mechanism for a new and an existing device to establish a secure out-of-band channel through which they can
communicate to facilitate the sign in of the new device.
```

reivilibre reviewed (commented):
not seeing anything materially wrong with this.
I did not review the cryptography very thoroughly but will trust that a cryptography review will take place.

hughns committed:
Apply suggestions from code review

Co-authored-by: Olivier 'reivilibre' <olivier@librepush.net>

poljar commented on proposals/4388-secure-qr-channel.md:
Well we can initialize it with some real exporter secret, but that would be more expensive and we wouldn't use the exporter interface of the response context.

Another alternative would be to not use the HPKE context for the other direction, but then we would need to reimplement the nonce handling.

hughns committed:
Bind secure channel to rendezvous session via additional authentication data on the HPKE Seal/Open invocations

hughns committed:
Ensure no leading zero in check code and notes on acceptable limitations

hughns committed:
Add alternative about making `Sec-Fetch` logic explicit

hughns committed:
With note about MUST and SHOULD

poljar commented on proposals/4388-secure-qr-channel.md:
You did not miss anything. It's well defined and clearly explained which sequence token is used when.

Thanks for adding this.

hughns committed:
Add discovery endpoint and remove unstable_features

hughns committed:
Add note about Context_DeviceG_Send exporter secret

hughns committed:
Attempt to clarify maximum size of data

hughns committed:
Clarify the PUT semantics

hughns committed:
Add create_available to discovery response body

hughns committed:
Correction on discovery response codes

uhoreg commented on proposals/4388-secure-qr-channel.md:
Note to other readers (including future-me, because I keep needing to remind myself how this works):

The security of this scheme is established by Device S decrypting and verifying this message.  The messages are encrypted by the result of the ECDH of G and S's ephemeral keys (plus extra steps).  S knows that it has G's public key because it was obtained from the QR scan (we assume that the attacker can't modify the displayed QR code).  So if it is able to decrypt and verify the message, then it knows:

- that it was received directly from G, since nobody other than S and G could obtain the shared secret to encrypt the message.  In other words, the message from G to S was not tampered with.
- that G has received the correct public key from S, otherwise the message from G would have been encrypted using the wrong key, and would be undecryptable/unverifiable.

Since, at this point, S knows that both devices have the correct public key for the other device, it knows that the encrypted channel is secure.

hughns committed:
Apply suggestions from review

Co-authored-by: Hubert Chathi <hubert@uhoreg.ca>

hughns committed:
Incorporate helpful note from @uhoreg

hughns committed:
Specify maximum length of rendezvous ID and sequence token + delimit AAD inputs

Resolves https://github.com/matrix-org/matrix-spec-proposals/pull/4388#discussion_r3025010846

hughns commented on proposals/4388-secure-qr-channel.md:
This is addressed by 5b6dc1eded7aa66cbd465469ad8a5b19fa6b883b

erikjohnston commented on proposals/4388-secure-qr-channel.md:
```diff
@@ -0,0 +1,1132 @@
+- Public/open - anyone can create a rendezvous session without an access token. This allows for the QR code to be
+  created on either the new or existing Matrix client.
+- Requires authenticated user - this would reduce abuse to known users, but would restrict the mechanism so that the QR
+  code must be created on the existing Matrix client (and therefore the new Matrix client must have a camera).
```
I think this is a bit misleading, as whether or not your server requires authentication doesn't change whether you can use QR on a new Matrix client? The new Matrix client will choose a well-known server to use as the rendezvous channel, no?


```suggestion
- Requires authenticated user - this would reduce abuse to known users. This should be the default.
- Public/open - anyone can create a rendezvous session without an access token. This allows for clients to use this
  server as its default when creating QR codes as a new session (as it doesn't yet know what server the user will
  log in to).
```

erikjohnston commented on proposals/4388-secure-qr-channel.md:
```diff
@@ -0,0 +1,1132 @@
+
+The server MUST perform a compare-and-swap operation by checking that the `sequence_token` matches
+the current sequence token for the session. If it does not match then the `data` MUST not be
+accepted and the `M_CONCURRENT_WRITE` error is returned.
```
Would be good to expand on how idempotency works here. Do we need to say the client MUST send the same payload with a given `sequence_token`? Presumably the server needs to check the payloads match if they get multiple requests to see if its a concurrent write or idempotency?

hughns replied:
Currently it isn't idempotent. I will make it so.

hughns replied:
Added in 98d3b04bf3841d82f1571073d8eec7d141146965. Synapse implementation in https://github.com/element-hq/synapse/pull/19808

erikjohnston commented on proposals/4388-secure-qr-channel.md:
```diff
@@ -0,0 +1,1132 @@
+
+The server MUST perform a compare-and-swap operation by checking that the `sequence_token` matches
+the current sequence token for the session. If it does not match then the `data` MUST not be
+accepted and the `M_CONCURRENT_WRITE` error is returned.
```
Would also be good to say what the client should do when they receive a `M_CONCURRENT_WRITE`, which is to do a `GET` to fetch the new payload and then retry?

hughns replied:
Incorporated in 9aabff85bfe8c97a338e9dc6e075b8c3b3f42d46

erikjohnston commented on proposals/4388-secure-qr-channel.md:
```diff
@@ -0,0 +1,1132 @@
+Every send request MUST include a `sequence_token` value whose value is the `sequence_token` from the last `GET`
+response seen by the requester. (The initiating device may also use the `sequence_token` supplied in the initial `POST` response
+to immediately update the payload.) Sends will succeed only if the supplied `sequence_token` matches the server's current
+revision of the payload. This prevents concurrent writes to the payload.
```
Might be worth spelling out a little that this isn't so much a channel that two clients can use to send a sequence of messages between them, but instead a shared mutable spot. This then makes the APIs below easier to reason about.

This also has consequences for how to use the session, since a client can't tell if their previous payload has been received by the remote client until a remote sets a payload itself and the original client sees the change. This favours an architecture where the protocol is "ping-pong", i.e. a client sends a message, waits for a reply, then can send the next message, etc. Or to send multiple messages in a row a client would update the payload to contain a list of messages.

hughns replied:
Incorporated in d125239e087b2ca90a4fb38c16d785c252ce5957

erikjohnston commented on proposals/4388-secure-qr-channel.md:
```diff
@@ -0,0 +1,1132 @@
+- `404 Not Found` (`M_NOT_FOUND`) - rendezvous session ID is not valid (it could have expired)
+- `404 Not Found` (`M_UNRECOGNIZED`) - the rendezvous API is not enabled
+- `429 Too Many Requests` (`M_LIMIT_EXCEEDED`) - the request has been rate limited
+
```
```suggestion

Cancelling a session will cause all future reads and writes to fail with a `404 M_NOT_FOUND`.

```

1vel commented on proposals/4388-secure-qr-channel.md:
Perhaps too late, but:
```suggestion
The unstable value of `IO.ELEMENT.MSC4388.CONCURRENT_WRITE` should be used instead of `M_CONCURRENT_WRITE`.
```

hughns replied:
Based on https://spec.matrix.org/v1.18/client-server-api/#standard-error-response it should actually be `IO.ELEMENT.MSC4388_CONCURRENT_WRITE` as the custom namespace is `io.element.msc4388`.

But, at this point I don't think it matters - we should focus on getting the stable names right 👍

hughns committed:
Update proposals/4388-secure-qr-channel.md

Co-authored-by: Erik Johnston <erik@matrix.org>

richvdh commented on proposals/4388-secure-qr-channel.md:
```diff
@@ -0,0 +1,1134 @@
+
+|Field|Type||
+|-|-|-|
+|`id`|required `string`|Opaque identifier for the rendezvous session. Maximum length 65,535 bytes|
```
64K sounds rather big for an opaque id, and particularly jarring given the limit for the payload of 4K. Could we use the standard opaque token grammar (https://spec.matrix.org/v1.17/appendices/#opaque-identifiers), or is 255 characters too short for cryptographic safety?

If not, you'll need to specify more detail about the allowed grammar. Are we talking Unicode strings (which then get encoded to UTF-8 by the JSON/REST API)? If so, please specify which characters are permitted (NUL? Unicode non-characters?). And the 64K limit applies to the UTF-8 encoding thereof?

hughns replied:
255 bytes would be fine and so have changed this to use the opaque identifier grammar in 914ae870b1c59b219947adbae40df7a233822642.

In the QR format the extra size byte has been removed to keep the QR as compact as practical. However, the AAD encoding is left as-is with the redundant size bytes considered to have little processing impact.

hughns replied:
Actually, on further discussion we have changed the AAD encoding to use two variants of `EncodeStringAsBytes()` in 914ae870b1c59b219947adbae40df7a233822642

richvdh commented on proposals/4388-secure-qr-channel.md:
```diff
@@ -0,0 +1,1134 @@
+|Field|Type||
+|-|-|-|
+|`id`|required `string`|Opaque identifier for the rendezvous session. Maximum length 65,535 bytes|
+|`sequence_token`|required `string`|The opaque token to identify if the payload has changed. Maximum length 65,535 bytes|
```
same goes for sequence_token I guess

hughns replied:
Done in 914ae870b1c59b219947adbae40df7a233822642

hughns committed:
Update proposals/4388-secure-qr-channel.md

Co-authored-by: Erik Johnston <erik@matrix.org>

hughns committed:
Incorporate review feedback

Co-Authored-By: Erik Johnston <github@jki.re>

hughns committed:
Add note about client handling of M_CONCURRENT_WRITE

hughns committed:
Make PUT requests idempotent

richvdh commented on proposals/4388-secure-qr-channel.md:
```diff
@@ -0,0 +1,1149 @@
+HTTP response codes:
+
+- `200 OK` - rendezvous session cancelled
+- `404 Not Found` (`M_NOT_FOUND`) - rendezvous session ID is not valid (it could have expired)
```
Since this seems to come up every time we discuss DELETE APIs:


```suggestion
- `404 Not Found` (`M_NOT_FOUND`) - rendezvous session ID is not valid (it could have expired,
  or been cancelled by a previous `DELETE` request).
```

richvdh commented on proposals/4388-secure-qr-channel.md:
```diff
@@ -0,0 +1,1149 @@
+
+#### `sequence_token` values
+
+The `sequence_token` values should be unique to the rendezvous session and the last modified time so that two clients can
```
Not quite sure how to read "unique to the rendezvous session". That sounds like a given server isn't allowed to use the same sequence token across different rendezvous sessions? I'm not sure why that would be necessary, but could you clarify?

hughns replied:
I have been back over my notes and I cannot find a reason for this requirement. I've checked the predecessor in MSC3886 to no avail.

In the secure channel layer we are already including both the rendezvous session ID and the sequence token in the additional authentication data calcuations. Therefore, I don't think we need the session ID to contribute to the sequence token at the lower level either.

As such I have removed in 64ff8a1f8633f6ef6f7ef3ed0cb2a8eddbdb4190

richvdh commented on proposals/4388-secure-qr-channel.md:
```diff
@@ -0,0 +1,1149 @@
+The "QR code version" would be repurposed to be a "QR code type" and used as the way to distinguish the format of the
+subsequent data.
+
+The existing cross verification code would be type `0x02`.  I suspect that type `0x01` and `0x00` might correspond to
+earlier iterations of the cross signing flow and so might want to be "reserved".
```
@uhoreg might remember the answer to this. 0x02 came from MSC1543, but I don't see any discussion of why 0x02.

uhoreg replied:
At least `0x01` was used for an earlier iteration.  But I don't remember if `0x00` was as well, or if I started counting at 1.

richvdh commented on proposals/4388-secure-qr-channel.md:
```suggestion
with an empty payload. It parses the response from the homeserver to extract the rendezvous session **ID**
and **sequence token**.
```

richvdh commented on proposals/4388-secure-qr-channel.md:
I'm finding this formatting all a bit hard to read and follow. Unfortunately Github can't cope with code blocks in suggestions, so have my suggestions in two parts:

```suggestion
Device S performs an ECDH operation using **Ss** and **Gp** to compute the shared secret **SharedSecret**. It then discards **Ss**.

The **SharedSecret** is used to initialize an
[HPKE encryption context](https://www.rfc-editor.org/rfc/rfc9180.html#name-encryption-and-decryption)
**Context_DeviceS_Send** using the `KeySchedule<ROLE>()` function, as follows:
```

hughns replied:
Ditto c56d9fa8fbc960a4610ecd167a83aa9429ffc4e3

richvdh commented on proposals/4388-secure-qr-channel.md:
```suggestion
With this, Device S has established its sending side of the secure channel. It then derives a confirmation
payload, **LoginInitiateMessage**, that Device G can use to confirm that the channel is secure. It contains:
```

richvdh commented on proposals/4388-secure-qr-channel.md:
The block below suggests that `Sp` comes first?

hughns replied:
Reordered in 98d3b04bf3841d82f1571073d8eec7d141146965

richvdh commented on proposals/4388-secure-qr-channel.md:
again, if Sp comes first in the message, this feels misleading


```suggestion
(Sp, TaggedCiphertext) := Unpack(LoginInitiateMessage)
```

richvdh commented on proposals/4388-secure-qr-channel.md:
```suggestion
AeadNonce := HKDF_SHA256(Secret, "nonce", salt=Salt, size=12)
```

richvdh commented on proposals/4388-secure-qr-channel.md:
I wondered what `[0;32]` was meant to mean. One possible way to clarify it:

```suggestion
It then derives its own HPKE context **Context_DeviceG_Send** for sending based on the scheme from the
[Oblivious HTTP](https://www.rfc-editor.org/rfc/rfc9458#name-encapsulation-of-responses) RFC, using a secret exported
from **Context_DeviceG_Receive**, and a dummy exporter secret, as follows:
```

... with some rewording of the warning after the code block.

hughns replied:
Assuming I've understood correctly then I have applied in c56d9fa8fbc960a4610ecd167a83aa9429ffc4e3

richvdh commented on proposals/4388-secure-qr-channel.md:
```suggestion
Device G then responds with a dummy payload, **LoginOkMessage**, containing the
```

richvdh commented on proposals/4388-secure-qr-channel.md:
Once more, assuming the `ResponseNonce` comes first as implied by the previous section, this would be more clearly written as:


```suggestion
(ResponseNonce, TaggedCiphertext) := Unpack(LoginOkMessage)
```

richvdh commented on proposals/4388-secure-qr-channel.md:
```suggestion
It then decrypts (and authenticates) the tagged ciphertext from the `LoginOkMessage` using the **Context_DeviceS_Receive** context, which will succeed provided
```

richvdh commented on proposals/4388-secure-qr-channel.md:
```suggestion
If the above was successful, Device S then calculates a two digit **CheckCode** code in the range [10, 99] using the [HPKE export
```

richvdh commented on proposals/4388-secure-qr-channel.md:
I note that `SequenceToken` is under control of the attacker in our threat model, making me wonder if we need to guard more carefully against replay attacks. I think the channel establishment messages defined in this MSC are probably fine, since they have a defined sequence and each embeds a message identifier in the plaintext, but is that true of MSC4108?

poljar replied:
More carefully compared to what's already described in the `Replay protection` [section](https://github.com/matrix-org/matrix-spec-proposals/pull/4388/changes#diff-cea6a6c22bb49f55c9626ed95c1dd720ba0bb909e72ff672f93cd944b7ef4b4dR949-R954) of this MSC and the [HPKE RCF](https://www.rfc-editor.org/rfc/rfc9180.html#name-replay-protection)?

Do you have some concrete additional protection in mind?

> but is that true of [MSC4108](https://github.com/matrix-org/matrix-doc/pull/4108)?

Could you expand on what you mean here? 

I'm going to try to answer as best as I can in the meantime.

Since MSC4108 depends on this MSC, all the replay protections provided by HPKE (each message has a unique sequence number which is used to derive the per-message nonce) and this MSC itself will apply to MSC4108.

Furthermore, the `SequenceToken` is now bound to the HPKE-internal sequence number by putting it into the AAD as well.

richvdh replied:
sorry for the delay in answering this, it got lost in the general github spam.

Anyway basically I'd missed that the HPKE context has an internal sequence number. Thanks.

richvdh reviewed (commented):
👍 looks very sound. A few comments.

hughns committed:
Apply suggestions from code review

Co-authored-by: Richard van der Hoff <1389908+richvdh@users.noreply.github.com>

hughns committed:
Apply suggestion from @richvdh

Co-authored-by: Richard van der Hoff <1389908+richvdh@users.noreply.github.com>

hughns committed:
Ordering of LoginInitiateMessage contents

hughns committed:
Readability improvements

Co-Authored-By: Richard van der Hoff <1389908+richvdh@users.noreply.github.com>

hughns committed:
Constrain rendezvous ID and sequence_token to the opaque identifier grammar

https://spec.matrix.org/v1.18/appendices/#opaque-identifiers

hughns committed:
Fix table inconsistency

hughns committed:
Replace EncodeStringAsBytes() with EncodeStringAsBytes8() and EncodeStringAsBytes16()

hughns committed:
sequence tokens do not need to be unique to a rendezvous session

richvdh commented on proposals/4388-secure-qr-channel.md:
```diff
@@ -413,7 +414,7 @@ The server MUST enforce a maximum `data` field size of 4096 bytes.
 #### `sequence_token` values
 
-The `sequence_token` values should be unique to the rendezvous session and the last modified time so that two clients can
+The `sequence_token` values should be unique to the last modified time so that two clients can
```
I'm not really sure what "unique to the last modified time" means, either. It opens the question of "at what time resolution?".  I think we just want to say that it must change each time the channel is written to.

We might want to be explicit that it must change even in the case of a no-op write (i.e. when writing the current value back.)

Do we have a suggested implementation? A simple counter?

hughns replied:
I've tried to address this in 726e115440a9e9125367b2bc23b690e1faec8787 by clarifying the expected behaviour (in the main section of the spec) and then having a recommended implementation (in the implementation notes section).

I will update the Synapse implementation to match.

hughns replied:
The Synapse implementation is in https://github.com/element-hq/synapse/pull/19906

richvdh commented on proposals/4388-secure-qr-channel.md:
As with a comment I made over at #4108, the formatting as `<ol>` lists makes for remarkably hard reading. Could you convert these to headings?

hughns replied:
Done in 3c267fd063090d3e0b260f2046cf13b56766e428

richvdh commented on proposals/4388-secure-qr-channel.md:
This bit doesn't seem to be "establishment", so needs a new heading. It seems to overlap with the "Secure operations" section below. Can we combine the two?

hughns replied:
Combined in 95e6b6d228e49835a9bc32e1144debad360c3b68

richvdh commented on proposals/4388-secure-qr-channel.md:
I'm not sure if this is unclear wording, or a hangover from a previous iteration of the protocol, but the per-message nonces come from the internals of the HPKE context. It would be good to clarify this.

hughns replied:
In 96239e7933fc88f9284deabd247bfb6ba35a98d1 I have deferred to the HPKE encryption contexts.

hughns committed:
<ol> to heading

hughns committed:
MSC4108 doesn't say where the check code is in the flow

MSC4388 is authoritative

hughns committed:
Refactor "secure operations" section

hughns committed:
Defer to HPKE for replay protection nonces

hughns committed:
Clarifications to sequence_token and recommended implementation

tulir commented on proposals/4388-secure-qr-channel.md:
```diff
@@ -0,0 +1,1169 @@
+
+Clients can use this endpoint to determine if the rendezvous API is available to them. Because the server policy may
+require authentication, clients should make this request with their access token if they have one.
+
+HTTP response codes, and Matrix error codes:
+
+- `200 OK` - rendezvous API discovery supported
+- `403 Forbidden` (`M_FORBIDDEN`) - the requester is not authorized to use the rendezvous API
+- `404 Not Found` (`M_UNRECOGNIZED`) - the rendezvous API is not enabled
+- `429 Too Many Requests` (`M_LIMIT_EXCEEDED`) - the request has been rate limited
+
+The response body for `200 OK` is `application/json` with contents:
+
+|Field|Type||
+|-|-|-|
+|`create_available`|required `boolean`|`true` if the requester is able to create a rendezvous session using `POST /_matrix/client/v1/rendezvous` otherwise `false`|
```
Is there any case where you'd return `M_FORBIDDEN` over `"create_available": false`? Seems a bit redundant

hughns replied:
What I recall wanting to achieve is allowing an existing authenticated client to distinguish between an expired access token (that can be refreshed) and the server policy disallowing the rendezvous from being created.

In the case of a refresh being required it looks like the correct `errcode` is actually `M_UNKNOWN_TOKEN` (not `M_FORBIDDEN`) and as it is a common error code it wouldn't need specifying here at all.

I will double check this and update.

tulir replied:
Yeah that sounds like it should be `M_UNKNOWN_TOKEN` and probably shouldn't be specified separately here. However, I'm not sure if there are any other endpoints that behave like this right now, so it might need some explanation.

Basically if auth is provided, but is invalid, then it'll fail the same way as an endpoint that requires auth would fail, even if auth isn't actually required

tulir commented on proposals/4388-secure-qr-channel.md:
```diff
@@ -0,0 +1,1169 @@
+
+#### `sequence_token` values
+
+A recommended implementation is a hash/digest of:
+
+- the rendezvous ID
+- a monotonic counter incremented on each successful write
+- the last `data` value written
```
Is there some practical benefit to this over just a monotonic counter?

hughns replied:
Looking at this afresh:

- the need to distinguish between identical payloads hails from ECIES was used to agree a shared symmetric key
- now that we are using two HPKE contexts the chances of having identical `data` in the payload are cryptographically small
- this sequence token was then (ab)used for the idempotency

A cleaner approach might be to:

- introduce an explicit client generated key that is sent with each `PUT` and is used to make those requests idempotent
- change the recommended sequence token format to be monotonic counter incremented on each successful write

Thoughts?

tulir replied:
Ah that explains it. A separate transaction ID does sound sensible, especially if the network is bad and one side is retrying PUT requests while the other side has already received the data and has overwritten it with a response. In the current proposal, that would result in a concurrent write error on the retries (because the active data no longer matches) even though the original write was successful. Server-side implementation of that would basically be a map of transaction ID to sequence token, which hopefully won't be too complicated or heavy (if input txn ID is in map, just return correspinding sequence ID).

tulir commented on proposals/4388-secure-qr-channel.md:
```suggestion
[opaque identifier grammar]) we don't specify a handling for `StringInput` of length greater than 255 bytes.
```

?

hughns replied:
Well spotted 👍

tulir commented on proposals/4388-secure-qr-channel.md:
```diff
@@ -0,0 +1,1169 @@
+  - the rendezvous session ID as a UTF-8 string
+- the [base URL] of the homeserver for client-server connections encoded as:
+  - two bytes in network byte order (big-endian) indicating the length in bytes of the base URL as a UTF-8 string
+    (n.b. a base URL longer than 65,535 bytes cannot be encoded and should be rejected)
```
Even 255 bytes would probably be more than enough for the base URL. DNS names can't be more than 253 bytes (and generally don't get anywhere close to the limit) and subpaths are very rare too

hughns committed:
Update proposals/4388-secure-qr-channel.md

Co-authored-by: Tulir Asokan <tulir@maunium.net>

turt2live reviewed (approved):
lgtm from when I've read it in the past
